doc/TACACSPLUS_PASSKEY_ENCRYPTION.md

```diff
@@ -0,0 +1,111 @@
+3. When TACACS+ Authentication is activated on the device, a new PAM configuration file (common-auth-sonic) is generated and substituted in the login and SSH daemons. Importantly, the pre-existing configuration file remains unchanged.
+
+The revised data handling procedure among the modules is outlined as follows:
+
+1. When a user configures the TACACS passkey using the SONIC CLI, it will now be securely stored in encrypted format instead of plaintext.
+2. Subsequently, the HostCfg Enforcer module retrieves this encrypted key. However, before writing it into the PAM configuration file(s), the hostCfgd module decrypts it. 
+```
+       +-------+  +---------+
+       |  SSH  |  | Console |
+       +---+---+  +----+----+
+           |           |   
++----------v-----------v---------+                                      +---------------------+
+| AUTHENTICATION                 |                                      |                     |
+|   +-------------------------+  |  Decrypted passkey                   |  +------------+     |
+|   | PAM Configuration Files <------------------------------------------+ |  AAA Config |    |
+|   +-------------------------+  |                                      |  +------------+     |
+|                                |                                      |                     |
+|         +-------------+        |                                      |  HostCfg Enforcer   |
+|         |     PAM     |        |                                      +----------^----------+
+|         |  Libraries  |        |                                                 |
+|         +-------------+        |                                                 | Encrypted passkey
++---------------+----------------+                                                 |
+                |                                                                  |
+           +----v----+                                                     +-------+--------+
+           |         |             Encrypted passkey                       |                |
+           |   CLI   +---------------------------------------------------->|    ConifgDB    |
+           |         |                                                     |                |
+           +---------+                                                     +----------------+
+```
+This decryption step is crucial because the login or SSH daemon references the PAM config file to verify the TACACS secret / passkey. If it remains encrypted, the SSH daemon will be unable to recognize the passkey, leading to login failures. The depicted block diagram clearly showcase the enhanced capbalities of the existing submodules.
+
+
+### Implementation details
+
+The implementation stands on three key pillars.
+1. OPENSSL toolkit is used for encryption / decryption
+2. aes-128-cbc is the encoding format used for encryption / decryption
+3. Device MAC address is used as a Password for encryption / decryption
```

Comment:
We can easily decrypt and know the actual password if we know the device MAC address, right?

Also, if I want to use the config_db (encrypted passkey) generated on one system to another (may be I forced to replace the failed node with same config), hope passkey cant be used on another device, what procedure should be followed for this use-case?

Reply:
Yes @venkatmahalingam  you are right. We can decrypt the passkey only if we knew the system MAC address. And if  we allow the user to use any custom password then it will be bit difficult in terms of maintainability. 

For instance, either we need to use some fixed cipher for encryption / decryption which will be more susceptible to the overall network breach. Alternatively,  we need to save the password somewhere (for instance in redis). This is the reason I chosen system MAC as a password which will make every encrypted passkey unique within the network.

An encrypted passkey on one node will not be applicable or useable on another as hostcfgd will not be able to decrypt that passkey. To handle this sort of scenario, the Orchestrator / network admin has to create the config_db.json or a encrypted passkey for every node (considering the orchestrator have the list of system MAC of all the nodes).